<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Simon Grimm">
<meta name="dcterms.date" content="2024-02-23">

<title>Simon’s Public NAO Notebook - Pooled swab sampling for pathogen early detection</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed fullcontent">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Simon’s Public NAO Notebook</span>
    </a>
  </div>
        <div class="quarto-navbar-tools">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Pooled swab sampling for pathogen early detection</h1>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Simon Grimm </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">February 23, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div id="b354ba1d" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="b354ba1d"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="b354ba1d-1"><a href="#b354ba1d-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Importing packages</span></span>
<span id="b354ba1d-2"><a href="#b354ba1d-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="b354ba1d-3"><a href="#b354ba1d-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="b354ba1d-4"><a href="#b354ba1d-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="b354ba1d-5"><a href="#b354ba1d-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="b354ba1d-6"><a href="#b354ba1d-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> collections <span class="im">import</span> defaultdict</span>
<span id="b354ba1d-7"><a href="#b354ba1d-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> random</span>
<span id="b354ba1d-8"><a href="#b354ba1d-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> math</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>The NAO aims to detect stealth pathogens at an early stage. Up until, we have have focused much our research on wastewater treatment plant (WTP) samples. While WTP sampling offers some advantages, its limitations include environmental noise, non-human biological material, and low pathogen abundance. The NAO thus also wants to investigate other sampling approaches. To this end we’ve previously researched the promise of air sampling for pathogen early detection, and Will recently created a framework that allows analysis of various types of sampling strategies.</p>
<p>One promising approach is sampling using nasal, nasopharyngeal (NP), or oropharyngeal (OP) swabs. When pooled, these samples have several advantages</p>
<ul>
<li>400x higher RA(1%) than wastewater, atleast based on an initial BOTEC, that I have low confidence in.</li>
<li>More metadata on individuals that contributed to the sample (because you can have them provide some basic metadata when providing the sample.</li>
<li>Potentially less static complexity than wastewater.</li>
<li>If pools are large enough, dynamic complexity could also be lower.</li>
<li>There is more existing research on swab sampling than on other sample types.</li>
</ul>
<p>Still, the approach also has downsides:</p>
<ul>
<li>If not sampling at an airport, getting wide geographic coverage is harder.</li>
<li>Scaling is more expensive than with wastewater, especially increasing the number of samples provided at at given location.</li>
<li>Regulatory hurdles are likely higher than when sampling wastewater.</li>
</ul>
</section>
<section id="pathogen-abundance-and-detection-capability" class="level1">
<h1>Pathogen abundance and detection capability</h1>
<section id="pathogen-shedding" class="level2">
<h2 class="anchored" data-anchor-id="pathogen-shedding">Pathogen shedding</h2>
<p>Among many types of pathogens, respiratory pathogens are particularly likely to cause future pandemics (Amesh A. Adalja, MD, Matthew Watson, …). Though different pathogens have different tropism across tissues (SARS-CoV-2 having a wider wider tropism than influenza (Flerlage et al.&nbsp;2021)), all of them are respiratory pathogens and thus shed in the respiratory tract, which includes the pharynx, mouth, and nose.</p>
<p>In a swab sampling program we could use different swabs that target any of these sitesm, i.e., the nostrils, the nasopharynx, oropharynx, and the mid-turbinate region<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a>. These sites differ in their comfort of sampling and (probably) in their sensitivity. For instance, nasopharyngeal swabs (NP swabs) would have be far better than other swabs for us to consider them for a sampling program, given that they do not allow self-sampling.</p>
<p>We will evaluate the performance of different swab types in three ways:</p>
<ul>
<li><strong>Sensitivity</strong>: Is the pathogen present at all?</li>
<li><strong>Viral load</strong>: How much of the pathogen can be measured? This metric is based on the qPCR cycle threshold of quantiative.</li>
<li><strong>Relative abundance</strong>: When performing untargeted sequencing, what is the relative abundance of a pathogen?</li>
</ul>
<section id="sensitivity-of-different-swab-types" class="level3">
<h3 class="anchored" data-anchor-id="sensitivity-of-different-swab-types">Sensitivity of different swab types</h3>
<p>(Tsang et al.&nbsp;2021) performed a meta-analysis to evaluate the ability of different swab types in detecting SARS-CoV-2 using qPCR. Swab types included nasal swabs (n=1622), throat swabs (n=388), and pooled nasal and throat swabs (n=719), each of which were compared to nasopharyngeal swabs. Overall, pooled nasal/throat swabs have the best diagnostic performance with sensitivity of 0.97 (0.93-1.00).</p>
<table class="table">
<caption><strong>Table 1: Comparison of different swab sample types.</strong> Adapted from (Tsang et al.&nbsp;2021). Values are effect sizes under a random effects-model. The gold standard is nasopharyngeal swabs.</caption>
<colgroup>
<col style="width: 25%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 23%">
<col style="width: 23%">
</colgroup>
<thead>
<tr class="header">
<th>Sample Type</th>
<th>Sensitivity</th>
<th>Specificity</th>
<th>Positive Predictive Value</th>
<th>Negative Predictive Value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Nasal swabs (n=1622)</td>
<td>0.86 (0.77-0.93)</td>
<td>0.99 (0.96-1.00)</td>
<td>0.96 (0.87-1.00)</td>
<td>0.95 (0.88-0.99)</td>
</tr>
<tr class="even">
<td>Throat swabs (n=388)</td>
<td>0.68 (0-35-0.94)</td>
<td>0.97 (0.95-0.99)</td>
<td>0.75 (0.45-0.96)</td>
<td>0.96 (0.94-0.98)</td>
</tr>
<tr class="odd">
<td>Pooled nasal/throat swabs (n=719)</td>
<td>0.97 (0.93-1.00)</td>
<td>0.99 (0.98-1.00)</td>
<td>0.97 (0.90-1.00)</td>
<td>0.99 (0.98-1.00)</td>
</tr>
</tbody>
</table>
<p>Note, though Tsang et al.&nbsp;2021 reports specificity, this metric isn’t very useful in this context. All studies covered by the review used qPCR to detect SARS-CoV-2. qPCR is generally considered to be very sensitive and specific. Thus, if a patient tests positive in a throat swab qPCR, but negative in nasopharyngeal swab qPCR, this shouldn’t be counted against throat swabs (false positive), but rather against nasopharyngeal swabs.</p>
<p>The review above gives us binary information about SARS-CoV-2 being present in different swab types. But we are not merely interested in a pathogen being present in a sample, but how abundant said pathogen is. For instance, hroat swabs coming back negative when NP swabs come back positive tells us that pathogen abundance is likely higher in the nasopharynx, but it’s unclear by how much.</p>
<p>Getting a better understanding of this is particularly relevant when pooling samples because higher relative abundance in an individual positive sample will ensure detection even if said sample is pooled with a large number of negative samples. To quantify this difference in absolute pathogen abundance between sample types we can look beyond positive/negative comparisons and instead look at the differences in CT scores within studies that compared swab types.</p>
</section>
<section id="viral-load-in-different-swab-types" class="level3">
<h3 class="anchored" data-anchor-id="viral-load-in-different-swab-types">Viral load in different swab types</h3>
<section id="nasopharyngeal-swabs" class="level4">
<h4 class="anchored" data-anchor-id="nasopharyngeal-swabs">Nasopharyngeal swabs</h4>
<p>Most studies on swab sampling performance treat nasopharyngeal swabs (NP swabs) as the gold standard, against which, nasal (Kojima et al.&nbsp;2021; Tu et al.&nbsp;2020), mid-turbinate (Tu et al.&nbsp;2020; Muller et al.&nbsp;2021), and combined nasal/oropharyngeal (Desmet et al.&nbsp;2021) swabs are compared. NP swabs are commonly administered by a healthcare professional and a properly administered test absorbs material from below the inferior turbinate, and the nasopharynx located at the back of the nasal cavity.</p>
<p>The CT value difference for paired NP/{Other} swabs are plotted in Figure 1. The CT value of the comparison swab is subtracted from the NP swab CT value. A lower CT is better, thus, a negative cycle threshold (CT) difference equates to a higher pathogen concentration in the NP swab.</p>
<div id="48b3ac5e" class="cell" data-execution_count="2">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="48b3ac5e"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="48b3ac5e-1"><a href="#48b3ac5e-1" aria-hidden="true" tabindex="-1"></a><span class="co"># All CT difference data were calculated in and are taken from https://docs.google.com/spreadsheets/d/1YP4mxT_vxiFwXU5ZuBM4obq_oscfhe05ODWwHFHEznM/</span></span>
<span id="48b3ac5e-2"><a href="#48b3ac5e-2" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.read_csv(<span class="st">"data/np_ct_differences.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="48b3ac5e-3"><a href="#48b3ac5e-3" aria-hidden="true" tabindex="-1"></a>df.columns <span class="op">=</span> df.columns.<span class="bu">str</span>.replace(<span class="st">", "</span>, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="48b3ac5e-4"><a href="#48b3ac5e-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="48b3ac5e-5"><a href="#48b3ac5e-5" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.melt(var_name<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>, value_name<span class="op">=</span><span class="st">"CT Difference"</span>)</span>
<span id="48b3ac5e-6"><a href="#48b3ac5e-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="48b3ac5e-7"><a href="#48b3ac5e-7" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="fl">3.5</span>))</span>
<span id="48b3ac5e-8"><a href="#48b3ac5e-8" aria-hidden="true" tabindex="-1"></a>sns.stripplot(</span>
<span id="48b3ac5e-9"><a href="#48b3ac5e-9" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df,</span>
<span id="48b3ac5e-10"><a href="#48b3ac5e-10" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="48b3ac5e-11"><a href="#48b3ac5e-11" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">"CT Difference"</span>,</span>
<span id="48b3ac5e-12"><a href="#48b3ac5e-12" aria-hidden="true" tabindex="-1"></a>    hue<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="48b3ac5e-13"><a href="#48b3ac5e-13" aria-hidden="true" tabindex="-1"></a>    jitter<span class="op">=</span><span class="va">True</span>,</span>
<span id="48b3ac5e-14"><a href="#48b3ac5e-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="48b3ac5e-15"><a href="#48b3ac5e-15" aria-hidden="true" tabindex="-1"></a>plt.legend([], [], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="48b3ac5e-16"><a href="#48b3ac5e-16" aria-hidden="true" tabindex="-1"></a><span class="co"># drop y axis label</span></span>
<span id="48b3ac5e-17"><a href="#48b3ac5e-17" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">""</span>)</span>
<span id="48b3ac5e-18"><a href="#48b3ac5e-18" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"qPCR Cycle Threshold Value Difference"</span>)</span>
<span id="48b3ac5e-19"><a href="#48b3ac5e-19" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">"y"</span>, which<span class="op">=</span><span class="st">"both"</span>, left<span class="op">=</span><span class="va">False</span>, right<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">True</span>)</span>
<span id="48b3ac5e-20"><a href="#48b3ac5e-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> <span class="dv">5</span>, <span class="dv">0</span>, <span class="op">-</span><span class="dv">5</span>, <span class="op">-</span><span class="dv">10</span>, <span class="op">-</span><span class="dv">15</span>:</span>
<span id="48b3ac5e-21"><a href="#48b3ac5e-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> x <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="48b3ac5e-22"><a href="#48b3ac5e-22" aria-hidden="true" tabindex="-1"></a>        plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"red"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="48b3ac5e-23"><a href="#48b3ac5e-23" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="48b3ac5e-24"><a href="#48b3ac5e-24" aria-hidden="true" tabindex="-1"></a>    plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="48b3ac5e-25"><a href="#48b3ac5e-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="48b3ac5e-26"><a href="#48b3ac5e-26" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> y <span class="kw">in</span> <span class="fl">0.5</span>, <span class="fl">1.5</span>, <span class="fl">2.5</span>, <span class="fl">3.5</span>, <span class="fl">4.5</span>:</span>
<span id="48b3ac5e-27"><a href="#48b3ac5e-27" aria-hidden="true" tabindex="-1"></a>    plt.axhline(y<span class="op">=</span>y, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="48b3ac5e-28"><a href="#48b3ac5e-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="48b3ac5e-29"><a href="#48b3ac5e-29" aria-hidden="true" tabindex="-1"></a>min_x, max_x <span class="op">=</span> plt.xlim()</span>
<span id="48b3ac5e-30"><a href="#48b3ac5e-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="48b3ac5e-31"><a href="#48b3ac5e-31" aria-hidden="true" tabindex="-1"></a>plt.text(max_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Comparison"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="48b3ac5e-32"><a href="#48b3ac5e-32" aria-hidden="true" tabindex="-1"></a>plt.text(min_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors NP Swab"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="48b3ac5e-33"><a href="#48b3ac5e-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="48b3ac5e-34"><a href="#48b3ac5e-34" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="48b3ac5e-35"><a href="#48b3ac5e-35" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="48b3ac5e-36"><a href="#48b3ac5e-36" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="48b3ac5e-37"><a href="#48b3ac5e-37" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="index_files/figure-html/cell-3-output-1.png" width="820" height="334" class="figure-img"></p>
<figcaption><strong>Figure 1: Nasopharyngeal vs Nasal, Mid-Turbinate, and combined Oropharyngeal/ Nasal swabs.</strong> Data from Kojima et al.&nbsp;2020, Tu et al.&nbsp;2020, Muller et al.&nbsp;2021, and Desmet et al.&nbsp;2021.</figcaption>
</figure>
</div>
</div>
</div>
<p>Apart from self-collected mid-turbinate swabs in Tu et al.&nbsp;2020, NP swabs show equivalent performance to nasal swabs in Kojima et al.&nbsp;2020, mid-turbinate swabs in Muller et al.&nbsp;2021, and combined nasal/throat swabs in Desmet et al 2021; and equivalent performance to nasal swabs in Tu et al.&nbsp;2020. In the same study, nasal swabs prove equivalent to NP swabs. Combined oro-pharyngeal swabs are also equivalent to NP swabs.</p>
<p>Studies that use NP swabs as their gold standard are useful to better understand the performance of other swab types, but NP swabs themselves are not practical for a large-scale pooled sampling programs: NP swabs are notoriously uncomfortable and are commonly administered by a third person. In contrast, nasal or oropharyngeal swabs would be more suitable for self-sampling, which allows far higher testing throughput. Let’s thus look at studies that compare nasal and oropharyngeal swabs.</p>
</section>
</section>
<section id="nasal-swabs" class="level3">
<h3 class="anchored" data-anchor-id="nasal-swabs">Nasal Swabs</h3>
<p>Note that the second plot shows the difference in genome copy numbers, rather than CT values<a href="#fn2" class="footnote-ref" id="fnref2" role="doc-noteref"><sup>2</sup></a>.</p>
<div id="536a0d53" class="cell" data-execution_count="3">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="536a0d53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="536a0d53-1"><a href="#536a0d53-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.read_csv(<span class="st">"data/nasal_ct_differences.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="536a0d53-2"><a href="#536a0d53-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-3"><a href="#536a0d53-3" aria-hidden="true" tabindex="-1"></a>df.columns <span class="op">=</span> df.columns.<span class="bu">str</span>.replace(<span class="st">", "</span>, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="536a0d53-4"><a href="#536a0d53-4" aria-hidden="true" tabindex="-1"></a><span class="co"># reshape dataframe to long format</span></span>
<span id="536a0d53-5"><a href="#536a0d53-5" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.melt(var_name<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>, value_name<span class="op">=</span><span class="st">"CT Difference"</span>)</span>
<span id="536a0d53-6"><a href="#536a0d53-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-7"><a href="#536a0d53-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-8"><a href="#536a0d53-8" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="fl">1.5</span>))</span>
<span id="536a0d53-9"><a href="#536a0d53-9" aria-hidden="true" tabindex="-1"></a>sns.stripplot(</span>
<span id="536a0d53-10"><a href="#536a0d53-10" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df,</span>
<span id="536a0d53-11"><a href="#536a0d53-11" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="536a0d53-12"><a href="#536a0d53-12" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">"CT Difference"</span>,</span>
<span id="536a0d53-13"><a href="#536a0d53-13" aria-hidden="true" tabindex="-1"></a>    hue<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="536a0d53-14"><a href="#536a0d53-14" aria-hidden="true" tabindex="-1"></a>    jitter<span class="op">=</span><span class="va">True</span>,</span>
<span id="536a0d53-15"><a href="#536a0d53-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="536a0d53-16"><a href="#536a0d53-16" aria-hidden="true" tabindex="-1"></a>plt.legend([], [], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="536a0d53-17"><a href="#536a0d53-17" aria-hidden="true" tabindex="-1"></a><span class="co"># drop y axis label</span></span>
<span id="536a0d53-18"><a href="#536a0d53-18" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">""</span>)</span>
<span id="536a0d53-19"><a href="#536a0d53-19" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"qPCR Cycle Threshold Value Difference"</span>)</span>
<span id="536a0d53-20"><a href="#536a0d53-20" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">"y"</span>, which<span class="op">=</span><span class="st">"both"</span>, left<span class="op">=</span><span class="va">False</span>, right<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">True</span>)</span>
<span id="536a0d53-21"><a href="#536a0d53-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-22"><a href="#536a0d53-22" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> <span class="dv">5</span>, <span class="dv">0</span>, <span class="op">-</span><span class="dv">5</span>, <span class="op">-</span><span class="dv">10</span>, <span class="op">-</span><span class="dv">15</span>:</span>
<span id="536a0d53-23"><a href="#536a0d53-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> x <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="536a0d53-24"><a href="#536a0d53-24" aria-hidden="true" tabindex="-1"></a>        plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"red"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="536a0d53-25"><a href="#536a0d53-25" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="536a0d53-26"><a href="#536a0d53-26" aria-hidden="true" tabindex="-1"></a>    plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="536a0d53-27"><a href="#536a0d53-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-28"><a href="#536a0d53-28" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="fl">0.5</span>, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="536a0d53-29"><a href="#536a0d53-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-30"><a href="#536a0d53-30" aria-hidden="true" tabindex="-1"></a>min_x, max_x <span class="op">=</span> plt.xlim()</span>
<span id="536a0d53-31"><a href="#536a0d53-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-32"><a href="#536a0d53-32" aria-hidden="true" tabindex="-1"></a>plt.text(max_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Comparison"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="536a0d53-33"><a href="#536a0d53-33" aria-hidden="true" tabindex="-1"></a>plt.text(min_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Nasal Swab"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="536a0d53-34"><a href="#536a0d53-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-35"><a href="#536a0d53-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-36"><a href="#536a0d53-36" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="536a0d53-37"><a href="#536a0d53-37" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="536a0d53-38"><a href="#536a0d53-38" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="536a0d53-39"><a href="#536a0d53-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="536a0d53-40"><a href="#536a0d53-40" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="index_files/figure-html/cell-4-output-1.png" width="820" height="186" class="figure-img"></p>
<figcaption><strong>Figure 2: Nasal swabs vs Oro-pharyngeal and combined nasal/oro-pharyngeal swabs.</strong> All data is taken from Goddall et al.&nbsp;2022.</figcaption>
</figure>
</div>
</div>
</div>
<div id="ee973545" class="cell" data-execution_count="4">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="ee973545"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="ee973545-1"><a href="#ee973545-1" aria-hidden="true" tabindex="-1"></a><span class="co"># All CT difference data was calculated and is taken from https://docs.google.com/spreadsheets/d/1YP4mxT_vxiFwXU5ZuBM4obq_oscfhe05ODWwHFHEznM/</span></span>
<span id="ee973545-2"><a href="#ee973545-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="ee973545-3"><a href="#ee973545-3" aria-hidden="true" tabindex="-1"></a><span class="co"># turn tsv into dataframe. Ignore first row. Second row is column names.</span></span>
<span id="ee973545-4"><a href="#ee973545-4" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.read_csv(<span class="st">"data/leung_genome_copy_differences.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="ee973545-5"><a href="#ee973545-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="ee973545-6"><a href="#ee973545-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="ee973545-7"><a href="#ee973545-7" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.melt(var_name<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>, value_name<span class="op">=</span><span class="st">"Genome Copy Number Difference"</span>)</span>
<span id="ee973545-8"><a href="#ee973545-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="ee973545-9"><a href="#ee973545-9" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"Study &amp; Comparison"</span>] <span class="op">=</span> df[<span class="st">"Study &amp; Comparison"</span>].<span class="bu">str</span>.split(<span class="st">","</span>).<span class="bu">str</span>[<span class="op">-</span><span class="dv">1</span>]</span>
<span id="ee973545-10"><a href="#ee973545-10" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">2</span>))</span>
<span id="ee973545-11"><a href="#ee973545-11" aria-hidden="true" tabindex="-1"></a>sns.stripplot(</span>
<span id="ee973545-12"><a href="#ee973545-12" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df,</span>
<span id="ee973545-13"><a href="#ee973545-13" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="ee973545-14"><a href="#ee973545-14" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">"Genome Copy Number Difference"</span>,</span>
<span id="ee973545-15"><a href="#ee973545-15" aria-hidden="true" tabindex="-1"></a>    hue<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="ee973545-16"><a href="#ee973545-16" aria-hidden="true" tabindex="-1"></a>    jitter<span class="op">=</span><span class="va">True</span>,</span>
<span id="ee973545-17"><a href="#ee973545-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="ee973545-18"><a href="#ee973545-18" aria-hidden="true" tabindex="-1"></a>plt.legend([], [], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="ee973545-19"><a href="#ee973545-19" aria-hidden="true" tabindex="-1"></a><span class="co"># drop y axis label</span></span>
<span id="ee973545-20"><a href="#ee973545-20" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">""</span>)</span>
<span id="ee973545-21"><a href="#ee973545-21" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Genome Copy Number Difference (Nasal - Throat Swabs), log10 scale"</span>)</span>
<span id="ee973545-22"><a href="#ee973545-22" aria-hidden="true" tabindex="-1"></a><span class="co"># flip x axis</span></span>
<span id="ee973545-23"><a href="#ee973545-23" aria-hidden="true" tabindex="-1"></a>plt.gca().invert_xaxis()</span>
<span id="ee973545-24"><a href="#ee973545-24" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">"y"</span>, which<span class="op">=</span><span class="st">"both"</span>, left<span class="op">=</span><span class="va">False</span>, right<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">True</span>)</span>
<span id="ee973545-25"><a href="#ee973545-25" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> <span class="op">-</span><span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>:</span>
<span id="ee973545-26"><a href="#ee973545-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> x <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="ee973545-27"><a href="#ee973545-27" aria-hidden="true" tabindex="-1"></a>        plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"red"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="ee973545-28"><a href="#ee973545-28" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="ee973545-29"><a href="#ee973545-29" aria-hidden="true" tabindex="-1"></a>    plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="ee973545-30"><a href="#ee973545-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="ee973545-31"><a href="#ee973545-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> y <span class="kw">in</span> <span class="fl">0.5</span>, <span class="fl">1.5</span>:</span>
<span id="ee973545-32"><a href="#ee973545-32" aria-hidden="true" tabindex="-1"></a>    plt.axhline(y<span class="op">=</span>y, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="ee973545-33"><a href="#ee973545-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="ee973545-34"><a href="#ee973545-34" aria-hidden="true" tabindex="-1"></a>min_x, max_x <span class="op">=</span> plt.xlim()</span>
<span id="ee973545-35"><a href="#ee973545-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="ee973545-36"><a href="#ee973545-36" aria-hidden="true" tabindex="-1"></a>plt.text(max_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Throat Swab"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="ee973545-37"><a href="#ee973545-37" aria-hidden="true" tabindex="-1"></a>plt.text(min_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Nasal Swab"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="ee973545-38"><a href="#ee973545-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="ee973545-39"><a href="#ee973545-39" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="ee973545-40"><a href="#ee973545-40" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="ee973545-41"><a href="#ee973545-41" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="ee973545-42"><a href="#ee973545-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="ee973545-43"><a href="#ee973545-43" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="index_files/figure-html/cell-5-output-1.png" width="760" height="223" class="figure-img"></p>
<figcaption><strong>Figure 3: Nasal swabs vs Oro-pharyngeal swabs.</strong> All data is taken from Leung et al.&nbsp;2020.</figcaption>
</figure>
</div>
</div>
</div>
<p>In qPCR measurements, nasal swabs contain more virus copies than throat swabs. In (Goodall et al.&nbsp;2022), nasal swabs are slightly superior to throat swabs and combined/nasal throat swabs, and come out about even when compared to combined nasal/OP swabs. In (Leung et al.&nbsp;2020), researchers ran multiplex-PCR on both nasal and throat swamples. qPCR differences are plotted in Figure 3 for Human Coronavirus, Influenza, and Rhinovirus.</p>
</section>
</section>
</section>
<section id="virus-relative-abundance-across-different-swab-types" class="level1">
<h1>Virus relative abundance across different swab types</h1>
<p>Whatever the sample type, ultimately we will want to perform metagenomic sequencing to detect novel pathogens. As described in a <a href="https://naobservatory.org/reports/comparing-sampling-strategies-for-early-detection-of-stealth-biothreats/">previous report</a>, the value of a metagenomic sequencing sample will can be assessed across several dimensions:</p>
<ul>
<li>What is the expected relative abundance of a human-infecting stealth virus?<a href="#fn3" class="footnote-ref" id="fnref3" role="doc-noteref"><sup>3</sup></a></li>
<li>How complex are individual sequencing samples? I.e, what is their species richness and evenness?</li>
<li>How much does the background complexity of a sample vary between individuals and over time?</li>
</ul>
<p>The relative abundance of a target pathogen can be impacted by its microbial background, because the sequencing efficiency of a taxon is directly dependant on the sampling efficiency of the other taxa in the sample<a href="#fn4" class="footnote-ref" id="fnref4" role="doc-noteref"><sup>4</sup></a></p>
<p>We identified two studies that performed paired metagenomic sequencing and SARS-CoV-2 qPCR on swab samples Lu et al.&nbsp;2021, Babiker et al.&nbsp;2020. We’ve reached out to the authors of two further studies that didn’t have publicly available data.</p>
<div id="6cc330e8" class="cell" data-execution_count="5">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="6cc330e8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="6cc330e8-1"><a href="#6cc330e8-1" aria-hidden="true" tabindex="-1"></a>df_throat <span class="op">=</span> pd.read_csv(<span class="st">"data/lu_throat_ct_mgs.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="6cc330e8-2"><a href="#6cc330e8-2" aria-hidden="true" tabindex="-1"></a>df_throat.rename(</span>
<span id="6cc330e8-3"><a href="#6cc330e8-3" aria-hidden="true" tabindex="-1"></a>    columns<span class="op">=</span>{<span class="st">"SCV-2 Relative Abundance"</span>: <span class="st">"scv2_ra"</span>, <span class="st">"Ct value"</span>: <span class="st">"scv2_ct"</span>}, inplace<span class="op">=</span><span class="va">True</span></span>
<span id="6cc330e8-4"><a href="#6cc330e8-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="6cc330e8-5"><a href="#6cc330e8-5" aria-hidden="true" tabindex="-1"></a>df_nasopharyngeal <span class="op">=</span> pd.read_csv(<span class="st">"data/babiker_np_ct_mgs.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="6cc330e8-6"><a href="#6cc330e8-6" aria-hidden="true" tabindex="-1"></a>df_nasopharyngeal.rename(</span>
<span id="6cc330e8-7"><a href="#6cc330e8-7" aria-hidden="true" tabindex="-1"></a>    columns<span class="op">=</span>{</span>
<span id="6cc330e8-8"><a href="#6cc330e8-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SARS-CoV-2 RT-PCR Ct"</span>: <span class="st">"scv2_ct"</span>,</span>
<span id="6cc330e8-9"><a href="#6cc330e8-9" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SARS-CoV-2 RA"</span>: <span class="st">"scv2_ra"</span>,</span>
<span id="6cc330e8-10"><a href="#6cc330e8-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Inpatient/ED vs. Outpatient"</span>: <span class="st">"patient_status"</span>,</span>
<span id="6cc330e8-11"><a href="#6cc330e8-11" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="6cc330e8-12"><a href="#6cc330e8-12" aria-hidden="true" tabindex="-1"></a>    inplace<span class="op">=</span><span class="va">True</span>,</span>
<span id="6cc330e8-13"><a href="#6cc330e8-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="6cc330e8-14"><a href="#6cc330e8-14" aria-hidden="true" tabindex="-1"></a>df_nasopharyngeal[<span class="st">"scv2_ct"</span>] <span class="op">=</span> (</span>
<span id="6cc330e8-15"><a href="#6cc330e8-15" aria-hidden="true" tabindex="-1"></a>    df_nasopharyngeal[<span class="st">"scv2_ct"</span>].replace(<span class="st">","</span>, <span class="st">"."</span>, regex<span class="op">=</span><span class="va">True</span>).astype(<span class="bu">float</span>)</span>
<span id="6cc330e8-16"><a href="#6cc330e8-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="6cc330e8-17"><a href="#6cc330e8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="6cc330e8-18"><a href="#6cc330e8-18" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">2</span>, <span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">9</span>, <span class="dv">6</span>))</span>
<span id="6cc330e8-19"><a href="#6cc330e8-19" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (ax, df) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(axs, [df_throat, df_nasopharyngeal])):</span>
<span id="6cc330e8-20"><a href="#6cc330e8-20" aria-hidden="true" tabindex="-1"></a>    x_lim <span class="op">=</span> <span class="fl">12.5</span>, <span class="fl">37.5</span></span>
<span id="6cc330e8-21"><a href="#6cc330e8-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="6cc330e8-22"><a href="#6cc330e8-22" aria-hidden="true" tabindex="-1"></a>        <span class="co"># set dot colors</span></span>
<span id="6cc330e8-23"><a href="#6cc330e8-23" aria-hidden="true" tabindex="-1"></a>        sns_default_colors <span class="op">=</span> sns.color_palette(<span class="st">"tab10"</span>)</span>
<span id="6cc330e8-24"><a href="#6cc330e8-24" aria-hidden="true" tabindex="-1"></a>        sns.scatterplot(</span>
<span id="6cc330e8-25"><a href="#6cc330e8-25" aria-hidden="true" tabindex="-1"></a>            ax<span class="op">=</span>ax,</span>
<span id="6cc330e8-26"><a href="#6cc330e8-26" aria-hidden="true" tabindex="-1"></a>            data<span class="op">=</span>df,</span>
<span id="6cc330e8-27"><a href="#6cc330e8-27" aria-hidden="true" tabindex="-1"></a>            x<span class="op">=</span><span class="st">"scv2_ct"</span>,</span>
<span id="6cc330e8-28"><a href="#6cc330e8-28" aria-hidden="true" tabindex="-1"></a>            y<span class="op">=</span><span class="st">"scv2_ra"</span>,</span>
<span id="6cc330e8-29"><a href="#6cc330e8-29" aria-hidden="true" tabindex="-1"></a>            hue<span class="op">=</span><span class="st">"patient_status"</span>,</span>
<span id="6cc330e8-30"><a href="#6cc330e8-30" aria-hidden="true" tabindex="-1"></a>            style<span class="op">=</span><span class="st">"patient_status"</span>,</span>
<span id="6cc330e8-31"><a href="#6cc330e8-31" aria-hidden="true" tabindex="-1"></a>            s<span class="op">=</span><span class="dv">100</span>,</span>
<span id="6cc330e8-32"><a href="#6cc330e8-32" aria-hidden="true" tabindex="-1"></a>            palette<span class="op">=</span>sns_default_colors[<span class="dv">2</span>:<span class="dv">4</span>],</span>
<span id="6cc330e8-33"><a href="#6cc330e8-33" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="6cc330e8-34"><a href="#6cc330e8-34" aria-hidden="true" tabindex="-1"></a>        <span class="co"># set coordinates for legend</span></span>
<span id="6cc330e8-35"><a href="#6cc330e8-35" aria-hidden="true" tabindex="-1"></a>        coordinates <span class="op">=</span> (<span class="fl">0.6</span>, <span class="op">-</span><span class="fl">0.2</span>)</span>
<span id="6cc330e8-36"><a href="#6cc330e8-36" aria-hidden="true" tabindex="-1"></a>        ax.legend(title<span class="op">=</span><span class="st">""</span>, loc<span class="op">=</span><span class="st">"upper left"</span>)  <span class="co"># bbox_to_anchor=coordinates)</span></span>
<span id="6cc330e8-37"><a href="#6cc330e8-37" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">"SARS-CoV-2 qPCR CT Value"</span>)</span>
<span id="6cc330e8-38"><a href="#6cc330e8-38" aria-hidden="true" tabindex="-1"></a>        ax.set_xlim(x_lim)</span>
<span id="6cc330e8-39"><a href="#6cc330e8-39" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="6cc330e8-40"><a href="#6cc330e8-40" aria-hidden="true" tabindex="-1"></a>        sns.scatterplot(ax<span class="op">=</span>ax, data<span class="op">=</span>df, x<span class="op">=</span><span class="st">"scv2_ct"</span>, y<span class="op">=</span><span class="st">"scv2_ra"</span>, s<span class="op">=</span><span class="dv">100</span>)</span>
<span id="6cc330e8-41"><a href="#6cc330e8-41" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">""</span>)</span>
<span id="6cc330e8-42"><a href="#6cc330e8-42" aria-hidden="true" tabindex="-1"></a>        ax.set_xlim(x_lim)</span>
<span id="6cc330e8-43"><a href="#6cc330e8-43" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">"log"</span>)</span>
<span id="6cc330e8-44"><a href="#6cc330e8-44" aria-hidden="true" tabindex="-1"></a>    ax.invert_xaxis()</span>
<span id="6cc330e8-45"><a href="#6cc330e8-45" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="6cc330e8-46"><a href="#6cc330e8-46" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="6cc330e8-47"><a href="#6cc330e8-47" aria-hidden="true" tabindex="-1"></a>    ax.tick_params(axis<span class="op">=</span><span class="st">"y"</span>, which<span class="op">=</span><span class="st">"both"</span>, left<span class="op">=</span><span class="va">False</span>, right<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">True</span>)</span>
<span id="6cc330e8-48"><a href="#6cc330e8-48" aria-hidden="true" tabindex="-1"></a>    ax.set_ylabel(<span class="st">"Relative Abundance"</span>)</span>
<span id="6cc330e8-49"><a href="#6cc330e8-49" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> x <span class="kw">in</span> np.arange(<span class="dv">15</span>, <span class="fl">37.5</span>, <span class="fl">2.5</span>):</span>
<span id="6cc330e8-50"><a href="#6cc330e8-50" aria-hidden="true" tabindex="-1"></a>        ax.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="6cc330e8-51"><a href="#6cc330e8-51" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> y <span class="kw">in</span> <span class="bu">range</span>(<span class="op">-</span><span class="dv">7</span>, <span class="dv">1</span>, <span class="dv">1</span>):</span>
<span id="6cc330e8-52"><a href="#6cc330e8-52" aria-hidden="true" tabindex="-1"></a>        log_y <span class="op">=</span> <span class="dv">10</span><span class="op">**</span>y</span>
<span id="6cc330e8-53"><a href="#6cc330e8-53" aria-hidden="true" tabindex="-1"></a>        ax.axhline(y<span class="op">=</span>log_y, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="index_files/figure-html/cell-6-output-1.png" width="748" height="503" class="figure-img"></p>
<figcaption><strong>Figure 4: SARS-CoV-2 qPCR CT vs SARS-CoV-2 RA.</strong> Top Figure: Throat Swabs. Data from Lu et al.&nbsp;2021. Bottom Figure: Nasopharyngeal Swabs. Data from Babiker et al.&nbsp;2020.</figcaption>
</figure>
</div>
</div>
</div>
<p>Some observations:</p>
<ul>
<li>Though comparing qPCR values between studies is tricky, we do see the expected difference in CT values, where nasopharyngeal swabs have lower CT values.</li>
<li>For the same CT value, Lu et al.&nbsp;2021 returns higher relative abundance.</li>
<li>Outpatients in Lu et al.&nbsp;have slightly higher relative abundances for the same CT value.</li>
</ul>
</section>
<section id="simulating-detection-capability" class="level1">
<h1>Simulating detection capability</h1>
<p>We can use the above data to replicate research by Jeff Kaufman. Let’s say we get a random set of swabs from a population which contains 1% sick people. On average 1% of all swabs will return positive, and the relative abundance in that positive sample is picked randomly among the relative abundances given in Lu et al.&nbsp;2021 and Babiker et al.&nbsp;2020, respectively.</p>
<p>Based on this, the median RA(1%) in a pooled sample with 200 swabs is ~ 10^-4.8 Lu et al.&nbsp;2021 (throat swabs) and 10^-5.9 in Babiker et al.&nbsp;2020 (NP swabs)</p>
<div id="f8e54b99" class="cell" data-execution_count="6">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="f8e54b99"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="f8e54b99-1"><a href="#f8e54b99-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> simulate_detection(positive_ras):</span>
<span id="f8e54b99-2"><a href="#f8e54b99-2" aria-hidden="true" tabindex="-1"></a>    FRAC_SICK <span class="op">=</span> <span class="fl">0.01</span></span>
<span id="f8e54b99-3"><a href="#f8e54b99-3" aria-hidden="true" tabindex="-1"></a>    SIMULATIONS <span class="op">=</span> <span class="dv">10000</span></span>
<span id="f8e54b99-4"><a href="#f8e54b99-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-5"><a href="#f8e54b99-5" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> probabilistic_round(x):</span>
<span id="f8e54b99-6"><a href="#f8e54b99-6" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="bu">int</span>(math.floor(x <span class="op">+</span> random.random()))</span>
<span id="f8e54b99-7"><a href="#f8e54b99-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-8"><a href="#f8e54b99-8" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> simulate_once(n_swabs):</span>
<span id="f8e54b99-9"><a href="#f8e54b99-9" aria-hidden="true" tabindex="-1"></a>        n_sick <span class="op">=</span> probabilistic_round(FRAC_SICK <span class="op">*</span> n_swabs)</span>
<span id="f8e54b99-10"><a href="#f8e54b99-10" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="kw">not</span> n_sick:</span>
<span id="f8e54b99-11"><a href="#f8e54b99-11" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> <span class="dv">0</span></span>
<span id="f8e54b99-12"><a href="#f8e54b99-12" aria-hidden="true" tabindex="-1"></a>        ra <span class="op">=</span> <span class="dv">0</span></span>
<span id="f8e54b99-13"><a href="#f8e54b99-13" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> _ <span class="kw">in</span> <span class="bu">range</span>(n_sick):</span>
<span id="f8e54b99-14"><a href="#f8e54b99-14" aria-hidden="true" tabindex="-1"></a>            ra <span class="op">=</span> random.choice(positive_ras)</span>
<span id="f8e54b99-15"><a href="#f8e54b99-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> ra <span class="op">/</span> n_swabs</span>
<span id="f8e54b99-16"><a href="#f8e54b99-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-17"><a href="#f8e54b99-17" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> simulate_many(n_simulations, n_swabs):</span>
<span id="f8e54b99-18"><a href="#f8e54b99-18" aria-hidden="true" tabindex="-1"></a>        RAs <span class="op">=</span> []</span>
<span id="f8e54b99-19"><a href="#f8e54b99-19" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> _ <span class="kw">in</span> <span class="bu">range</span>(n_simulations):</span>
<span id="f8e54b99-20"><a href="#f8e54b99-20" aria-hidden="true" tabindex="-1"></a>            RAs.append(simulate_once(n_swabs))</span>
<span id="f8e54b99-21"><a href="#f8e54b99-21" aria-hidden="true" tabindex="-1"></a>        RAs.sort()</span>
<span id="f8e54b99-22"><a href="#f8e54b99-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-23"><a href="#f8e54b99-23" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> [</span>
<span id="f8e54b99-24"><a href="#f8e54b99-24" aria-hidden="true" tabindex="-1"></a>            RAs[probabilistic_round(<span class="bu">len</span>(RAs) <span class="op">*</span> percentile <span class="op">/</span> <span class="dv">100</span>)]</span>
<span id="f8e54b99-25"><a href="#f8e54b99-25" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> percentile <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">100</span>)</span>
<span id="f8e54b99-26"><a href="#f8e54b99-26" aria-hidden="true" tabindex="-1"></a>        ]</span>
<span id="f8e54b99-27"><a href="#f8e54b99-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-28"><a href="#f8e54b99-28" aria-hidden="true" tabindex="-1"></a>    data <span class="op">=</span> defaultdict(<span class="bu">list</span>)</span>
<span id="f8e54b99-29"><a href="#f8e54b99-29" aria-hidden="true" tabindex="-1"></a>    n_swab_range <span class="op">=</span> [<span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">400</span>, <span class="dv">800</span>]</span>
<span id="f8e54b99-30"><a href="#f8e54b99-30" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> n_swabs <span class="kw">in</span> n_swab_range:</span>
<span id="f8e54b99-31"><a href="#f8e54b99-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> percentile, ra <span class="kw">in</span> <span class="bu">enumerate</span>(simulate_many(SIMULATIONS, n_swabs)):</span>
<span id="f8e54b99-32"><a href="#f8e54b99-32" aria-hidden="true" tabindex="-1"></a>            data[percentile].append(ra)</span>
<span id="f8e54b99-33"><a href="#f8e54b99-33" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> pd.DataFrame(data).T</span>
<span id="f8e54b99-34"><a href="#f8e54b99-34" aria-hidden="true" tabindex="-1"></a>    df.columns <span class="op">=</span> n_swab_range</span>
<span id="f8e54b99-35"><a href="#f8e54b99-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-36"><a href="#f8e54b99-36" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> df</span>
<span id="f8e54b99-37"><a href="#f8e54b99-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-38"><a href="#f8e54b99-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-39"><a href="#f8e54b99-39" aria-hidden="true" tabindex="-1"></a>df_throat <span class="op">=</span> pd.read_csv(<span class="st">"data/lu_throat_ct_mgs.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="f8e54b99-40"><a href="#f8e54b99-40" aria-hidden="true" tabindex="-1"></a>df_throat.rename(</span>
<span id="f8e54b99-41"><a href="#f8e54b99-41" aria-hidden="true" tabindex="-1"></a>    columns<span class="op">=</span>{<span class="st">"SCV-2 Relative Abundance"</span>: <span class="st">"scv2_ra"</span>, <span class="st">"Ct value"</span>: <span class="st">"scv2_ct"</span>}, inplace<span class="op">=</span><span class="va">True</span></span>
<span id="f8e54b99-42"><a href="#f8e54b99-42" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="f8e54b99-43"><a href="#f8e54b99-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-44"><a href="#f8e54b99-44" aria-hidden="true" tabindex="-1"></a>df_nasopharyngeal <span class="op">=</span> pd.read_csv(<span class="st">"data/babiker_np_ct_mgs.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="f8e54b99-45"><a href="#f8e54b99-45" aria-hidden="true" tabindex="-1"></a>df_nasopharyngeal.rename(</span>
<span id="f8e54b99-46"><a href="#f8e54b99-46" aria-hidden="true" tabindex="-1"></a>    columns<span class="op">=</span>{</span>
<span id="f8e54b99-47"><a href="#f8e54b99-47" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SARS-CoV-2 RT-PCR Ct"</span>: <span class="st">"scv2_ct"</span>,</span>
<span id="f8e54b99-48"><a href="#f8e54b99-48" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SARS-CoV-2 RA"</span>: <span class="st">"scv2_ra"</span>,</span>
<span id="f8e54b99-49"><a href="#f8e54b99-49" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Inpatient/ED vs. Outpatient"</span>: <span class="st">"patient_status"</span>,</span>
<span id="f8e54b99-50"><a href="#f8e54b99-50" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="f8e54b99-51"><a href="#f8e54b99-51" aria-hidden="true" tabindex="-1"></a>    inplace<span class="op">=</span><span class="va">True</span>,</span>
<span id="f8e54b99-52"><a href="#f8e54b99-52" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="f8e54b99-53"><a href="#f8e54b99-53" aria-hidden="true" tabindex="-1"></a>df_nasopharyngeal[<span class="st">"scv2_ct"</span>] <span class="op">=</span> (</span>
<span id="f8e54b99-54"><a href="#f8e54b99-54" aria-hidden="true" tabindex="-1"></a>    df_throat[<span class="st">"scv2_ct"</span>].replace(<span class="st">","</span>, <span class="st">"."</span>, regex<span class="op">=</span><span class="va">True</span>).astype(<span class="bu">float</span>)</span>
<span id="f8e54b99-55"><a href="#f8e54b99-55" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="f8e54b99-56"><a href="#f8e54b99-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-57"><a href="#f8e54b99-57" aria-hidden="true" tabindex="-1"></a>df_throat_ras <span class="op">=</span> df_throat[<span class="st">"scv2_ra"</span>].dropna().tolist()</span>
<span id="f8e54b99-58"><a href="#f8e54b99-58" aria-hidden="true" tabindex="-1"></a>df_nasopharyngeal_ras <span class="op">=</span> df_nasopharyngeal[<span class="st">"scv2_ra"</span>].dropna().tolist()</span>
<span id="f8e54b99-59"><a href="#f8e54b99-59" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">9</span>, <span class="dv">5</span>), sharey<span class="op">=</span><span class="va">True</span>)</span>
<span id="f8e54b99-60"><a href="#f8e54b99-60" aria-hidden="true" tabindex="-1"></a>n_swab_range <span class="op">=</span> [<span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">400</span>, <span class="dv">800</span>]</span>
<span id="f8e54b99-61"><a href="#f8e54b99-61" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, positive_ras <span class="kw">in</span> <span class="bu">enumerate</span>([df_throat_ras, df_nasopharyngeal_ras]):</span>
<span id="f8e54b99-62"><a href="#f8e54b99-62" aria-hidden="true" tabindex="-1"></a>    df <span class="op">=</span> simulate_detection(positive_ras)</span>
<span id="f8e54b99-63"><a href="#f8e54b99-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> n_swabs <span class="kw">in</span> n_swab_range:</span>
<span id="f8e54b99-64"><a href="#f8e54b99-64" aria-hidden="true" tabindex="-1"></a>        axs[i].plot(df.index, df[n_swabs], label<span class="op">=</span><span class="ss">f"</span><span class="sc">{</span>n_swabs<span class="sc">}</span><span class="ss"> swabs"</span>)</span>
<span id="f8e54b99-65"><a href="#f8e54b99-65" aria-hidden="true" tabindex="-1"></a>    axs[i].set_xlabel(<span class="st">"Percentile"</span>)</span>
<span id="f8e54b99-66"><a href="#f8e54b99-66" aria-hidden="true" tabindex="-1"></a>    axs[i].set_ylabel(<span class="st">"RA"</span>)</span>
<span id="f8e54b99-67"><a href="#f8e54b99-67" aria-hidden="true" tabindex="-1"></a>    axs[i].set_yscale(<span class="st">"log"</span>)</span>
<span id="f8e54b99-68"><a href="#f8e54b99-68" aria-hidden="true" tabindex="-1"></a>    <span class="co"># turn on y tick labels for both plots</span></span>
<span id="f8e54b99-69"><a href="#f8e54b99-69" aria-hidden="true" tabindex="-1"></a>    axs[i].yaxis.set_tick_params(labelleft<span class="op">=</span><span class="va">True</span>)</span>
<span id="f8e54b99-70"><a href="#f8e54b99-70" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> x <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="dv">100</span>, <span class="dv">20</span>):</span>
<span id="f8e54b99-71"><a href="#f8e54b99-71" aria-hidden="true" tabindex="-1"></a>        axs[i].axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"black"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="f8e54b99-72"><a href="#f8e54b99-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-73"><a href="#f8e54b99-73" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> y <span class="kw">in</span> <span class="bu">range</span>(<span class="op">-</span><span class="dv">1</span>, <span class="op">-</span><span class="dv">10</span>, <span class="op">-</span><span class="dv">1</span>):</span>
<span id="f8e54b99-74"><a href="#f8e54b99-74" aria-hidden="true" tabindex="-1"></a>        log_y <span class="op">=</span> <span class="dv">10</span><span class="op">**</span>y</span>
<span id="f8e54b99-75"><a href="#f8e54b99-75" aria-hidden="true" tabindex="-1"></a>        axs[i].axhline(y<span class="op">=</span>log_y, color<span class="op">=</span><span class="st">"black"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="f8e54b99-76"><a href="#f8e54b99-76" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="f8e54b99-77"><a href="#f8e54b99-77" aria-hidden="true" tabindex="-1"></a>        axs[i].set_title(<span class="st">"Throat"</span>)</span>
<span id="f8e54b99-78"><a href="#f8e54b99-78" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="f8e54b99-79"><a href="#f8e54b99-79" aria-hidden="true" tabindex="-1"></a>        axs[i].set_title(<span class="st">"Nasopharyngeal"</span>)</span>
<span id="f8e54b99-80"><a href="#f8e54b99-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="f8e54b99-81"><a href="#f8e54b99-81" aria-hidden="true" tabindex="-1"></a>    axs[i].spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="f8e54b99-82"><a href="#f8e54b99-82" aria-hidden="true" tabindex="-1"></a>    axs[i].spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="f8e54b99-83"><a href="#f8e54b99-83" aria-hidden="true" tabindex="-1"></a>    ax.tick_params(axis<span class="op">=</span><span class="st">"x"</span>, which<span class="op">=</span><span class="st">"both"</span>, bottom<span class="op">=</span><span class="va">False</span>, top<span class="op">=</span><span class="va">False</span>, labelbottom<span class="op">=</span><span class="va">True</span>)</span>
<span id="f8e54b99-84"><a href="#f8e54b99-84" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="f8e54b99-85"><a href="#f8e54b99-85" aria-hidden="true" tabindex="-1"></a>        axs[i].legend()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="index_files/figure-html/cell-7-output-1.png" width="749" height="449" class="figure-img"></p>
<figcaption><strong>Figure 5: Simulated detection capability of pooled swab sampling.</strong> Left Figure: Throat Swabs. Data from Lu et al.&nbsp;2021. Right Figure: Nasopharyngeal Swabs. Data from Babiker et al.&nbsp;2020.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="supplement" class="level1">
<h1>Supplement</h1>
<p>Create a figure for the remaining viruses not covered in Figure 3.</p>


</section>


<div id="quarto-appendix" class="default"><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><h2 class="anchored quarto-appendix-heading">Footnotes</h2>

<ol>
<li id="fn1"><p>The shedding dynamics of a <strong>stealth pathogen</strong> are likely to be different from SARS-CoV-2, so we need to be careful about generalizing from samples that originate from SARS-CoV-2 patients, particularly if they are hospitalized.<a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn2"><p>A 1-log difference in genome copy number is equivalent to a 3.3 CT difference. This is because each cycle in a qPCR machine approxiately equates to doubling of the genome copy number.<a href="#fnref2" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn3"><p>The shedding dynamics of a <strong>stealth pathogen</strong> are likely to be different from SARS-CoV-2, so we need to be careful about generalizing from samples that originate from SARS-CoV-2 patients, particularly if they are hospitalized.<a href="#fnref3" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
<li id="fn4"><p>See <a href="https://elifesciences.org/articles/46923">McClaren et al.&nbsp;2019</a> for further discussion.<a href="#fnref4" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol>
</section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>