<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.550">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Simon Grimm">
<meta name="dcterms.date" content="2024-02-23">

<title>Pooled swab sampling for pathogen early detection</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="swab-sampling_files/libs/clipboard/clipboard.min.js"></script>
<script src="swab-sampling_files/libs/quarto-html/quarto.js"></script>
<script src="swab-sampling_files/libs/quarto-html/popper.min.js"></script>
<script src="swab-sampling_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="swab-sampling_files/libs/quarto-html/anchor.min.js"></script>
<link href="swab-sampling_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="swab-sampling_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="swab-sampling_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="swab-sampling_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="swab-sampling_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content page-columns page-full" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Pooled swab sampling for pathogen early detection</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Simon Grimm </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">February 23, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<div id="3359c48f" class="cell" data-execution_count="1">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="3359c48f"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="3359c48f-1"><a href="#3359c48f-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Importing packages</span></span>
<span id="3359c48f-2"><a href="#3359c48f-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="3359c48f-3"><a href="#3359c48f-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="3359c48f-4"><a href="#3359c48f-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="3359c48f-5"><a href="#3359c48f-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="introduction" class="level1">
<h1>Introduction</h1>
<p>The NAO aims to detect stealth pathogens at an early stage. While WTP sampling offers some advantages, its limitations include environmental noise, non-human biological material, and low pathogen abundance. The NAO thus also wants to investigate other sampling approaches. To this end we’ve previously researched the promise of air sampling for pathogen early detection, and Will recently created a framework that allows analysis of various types of sampling strategies.</p>
<p>One promising approach is sampling using nasal, nasopharyngeal (NP), or oropharyngeal (OP) swabs. When pooled, these samples have several advantages [high relative abundance, more metadata, probably less dynamic complexity than wastewater, probably less static complexity than wastewater]. It also has downsides [difficult to get samples with large geographic coverage, scaling is hard, EGD is harder, regulatory hurdles are higher than with wastewater].</p>
</section>
<section id="pathogen-abundance-and-detection-capability" class="level1 page-columns page-full">
<h1>Pathogen abundance and detection capability</h1>
<section id="pathogen-shedding" class="level2">
<h2 class="anchored" data-anchor-id="pathogen-shedding">Pathogen shedding</h2>
<p>Respiratory pathogens are likely to cause future pandemics (Amesh A. Adalja, MD, Matthew Watson, …) Though the tropism across different tissues differs across respiratory pathogens (SARS-CoV-2 has a wider tropism than influenza (Flerlage et al.&nbsp;2021)), all of them are likely to shed predominantly in the respiratory tract, which includes the pharynx, mouth, and nose.</p>
<p>Sensitivity (Tsang et al.&nbsp;2021) performed a meta-analysis to evaluate the ability of different swab types in detecting SARS-CoV-2 using qPCR. Swab types included nasal swabs (n=1622), throat swabs (n=388), and pooled nasal and throat swabs (n=719), each of which were compared to nasopharyngeal swabs. Overall, pooled nasal/throat swabs have the best diagnostic performance with sensitivity of 0.97 (0.93-1.00).</p>
<table class="table">
<colgroup>
<col style="width: 25%">
<col style="width: 13%">
<col style="width: 13%">
<col style="width: 23%">
<col style="width: 23%">
</colgroup>
<thead>
<tr class="header">
<th>Sample Type</th>
<th>Sensitivity</th>
<th>Specificity</th>
<th>Positive Predictive Value</th>
<th>Negative Predictive Value</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Nasal swabs (n=1622)</td>
<td>0.86 (0.77-0.93)</td>
<td>0.99 (0.96-1.00)</td>
<td>0.96 (0.87-1.00)</td>
<td>0.95 (0.88-0.99)</td>
</tr>
<tr class="even">
<td>Throat swabs (n=388)</td>
<td>0.68 (0-35-0.94)</td>
<td>0.97 (0.95-0.99)</td>
<td>0.75 (0.45-0.96)</td>
<td>0.96 (0.94-0.98)</td>
</tr>
<tr class="odd">
<td>Pooled nasal/throat swabs (n=719)</td>
<td>0.97 (0.93-1.00)</td>
<td>0.99 (0.98-1.00)</td>
<td>0.97 (0.90-1.00)</td>
<td>0.99 (0.98-1.00)</td>
</tr>
</tbody>
</table>
<p>Table X: Comparison of different swab sample types. Adapted from (Tsang et al.&nbsp;2021). Values are effect sizes under a random effects-model. The gold standard is nasopharyngeal swabs.</p>
<p>Though Tsang et al.&nbsp;2021 reports specificity, this metric isn’t very useful in this context. All studies covered by the review used qPCR to detect SARS-CoV-2. qPCR is generally considered to be very sensitive and specific. Thus, if a patient tests positive in a throat swab qPCR, but negative in nasopharyngeal swab qPCR, this shouldn’t be counted against throat swabs (false positive), but rather against nasopharyngeal swabs.</p>
<p>The review above gives us binary information about SARS-CoV-2 being present in different swab types. But we are not merely interested in a pathogen being present in a sample at all, but how abundant said pathogen is. E.g., throat swabs coming back negative when NP swabs come back positive tells us that pathogen abundance is likely higher in the nasopharynx, but it’s unclear by how much. Getting a better understanding of this is particularly relevant when pooling samples, as higher relative abundance in a positive sample will ensure detection even if said sample is pooled with a large number of negative samples. To quantify this difference in absolute pathogen abundance between sample types we can look beyond positive/negative comparisons and instead look at the differences in CT scores within studies that made sample type comparisons.</p>
</section>
<section id="viral-load-in-different-swab-types" class="level2 page-columns page-full">
<h2 class="anchored" data-anchor-id="viral-load-in-different-swab-types">Viral load in different swab types</h2>
<section id="nasopharyngeal-swabs" class="level3 page-columns page-full">
<h3 class="anchored" data-anchor-id="nasopharyngeal-swabs">Nasopharyngeal swabs</h3>
<p>Most studies on swab sampling performance treat nasopharyngeal swabs (NP swabs) as the gold standard, against which, nasal (Kojima et al.&nbsp;2021; Tu et al.&nbsp;2020), mid-turbinate (Tu et al.&nbsp;2020; Muller et al.&nbsp;2021), and combined nasal/oropharyngeal (Desmet et al.&nbsp;2021) swabs are compared. NP swabs are commonly administered by a healthcare professional and a properly administered test absorbs material from below the inferior turbinate, and the nasopharynx located at the back of the nasal cavity.</p>
<p>The CT value difference for paired NP/{Other} swabs are plotted in Figure 1. The CT value of the comparison swab is subtracted from the NP swab CT value. A lower CT is better, thus, a negative cycle threshold (CT) difference equates to a higher pathogen concentration in the NP swab.</p>
<div id="a23a03aa" class="cell page-columns page-full" data-execution_count="2">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="a23a03aa"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="a23a03aa-1"><a href="#a23a03aa-1" aria-hidden="true" tabindex="-1"></a><span class="co"># All CT difference data were calculated in and are taken from https://docs.google.com/spreadsheets/d/1YP4mxT_vxiFwXU5ZuBM4obq_oscfhe05ODWwHFHEznM/</span></span>
<span id="a23a03aa-2"><a href="#a23a03aa-2" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.read_csv(<span class="st">"data/np_ct_differences.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="a23a03aa-3"><a href="#a23a03aa-3" aria-hidden="true" tabindex="-1"></a>df.columns <span class="op">=</span> df.columns.<span class="bu">str</span>.replace(<span class="st">", "</span>, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="a23a03aa-4"><a href="#a23a03aa-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="a23a03aa-5"><a href="#a23a03aa-5" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.melt(var_name<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>, value_name<span class="op">=</span><span class="st">"CT Difference"</span>)</span>
<span id="a23a03aa-6"><a href="#a23a03aa-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="a23a03aa-7"><a href="#a23a03aa-7" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="fl">3.5</span>))</span>
<span id="a23a03aa-8"><a href="#a23a03aa-8" aria-hidden="true" tabindex="-1"></a>sns.stripplot(</span>
<span id="a23a03aa-9"><a href="#a23a03aa-9" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df,</span>
<span id="a23a03aa-10"><a href="#a23a03aa-10" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="a23a03aa-11"><a href="#a23a03aa-11" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">"CT Difference"</span>,</span>
<span id="a23a03aa-12"><a href="#a23a03aa-12" aria-hidden="true" tabindex="-1"></a>    hue<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="a23a03aa-13"><a href="#a23a03aa-13" aria-hidden="true" tabindex="-1"></a>    jitter<span class="op">=</span><span class="va">True</span>,</span>
<span id="a23a03aa-14"><a href="#a23a03aa-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="a23a03aa-15"><a href="#a23a03aa-15" aria-hidden="true" tabindex="-1"></a>plt.legend([], [], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="a23a03aa-16"><a href="#a23a03aa-16" aria-hidden="true" tabindex="-1"></a><span class="co"># drop y axis label</span></span>
<span id="a23a03aa-17"><a href="#a23a03aa-17" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">""</span>)</span>
<span id="a23a03aa-18"><a href="#a23a03aa-18" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"qPCR Cycle Threshold Value Difference"</span>)</span>
<span id="a23a03aa-19"><a href="#a23a03aa-19" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">"y"</span>, which<span class="op">=</span><span class="st">"both"</span>, left<span class="op">=</span><span class="va">False</span>, right<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">True</span>)</span>
<span id="a23a03aa-20"><a href="#a23a03aa-20" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> <span class="dv">5</span>, <span class="dv">0</span>, <span class="op">-</span><span class="dv">5</span>, <span class="op">-</span><span class="dv">10</span>, <span class="op">-</span><span class="dv">15</span>:</span>
<span id="a23a03aa-21"><a href="#a23a03aa-21" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> x <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="a23a03aa-22"><a href="#a23a03aa-22" aria-hidden="true" tabindex="-1"></a>        plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"red"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="a23a03aa-23"><a href="#a23a03aa-23" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="a23a03aa-24"><a href="#a23a03aa-24" aria-hidden="true" tabindex="-1"></a>    plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="a23a03aa-25"><a href="#a23a03aa-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="a23a03aa-26"><a href="#a23a03aa-26" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> y <span class="kw">in</span> <span class="fl">0.5</span>, <span class="fl">1.5</span>, <span class="fl">2.5</span>, <span class="fl">3.5</span>, <span class="fl">4.5</span>:</span>
<span id="a23a03aa-27"><a href="#a23a03aa-27" aria-hidden="true" tabindex="-1"></a>    plt.axhline(y<span class="op">=</span>y, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="a23a03aa-28"><a href="#a23a03aa-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="a23a03aa-29"><a href="#a23a03aa-29" aria-hidden="true" tabindex="-1"></a>min_x, max_x <span class="op">=</span> plt.xlim()</span>
<span id="a23a03aa-30"><a href="#a23a03aa-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="a23a03aa-31"><a href="#a23a03aa-31" aria-hidden="true" tabindex="-1"></a>plt.text(max_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Comparison"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="a23a03aa-32"><a href="#a23a03aa-32" aria-hidden="true" tabindex="-1"></a>plt.text(min_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors NP Swab"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="a23a03aa-33"><a href="#a23a03aa-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="a23a03aa-34"><a href="#a23a03aa-34" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="a23a03aa-35"><a href="#a23a03aa-35" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="a23a03aa-36"><a href="#a23a03aa-36" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="a23a03aa-37"><a href="#a23a03aa-37" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full">
<p><img src="swab-sampling_files/figure-html/cell-3-output-1.png" class="figure-img" width="820" height="334"></p>
<figcaption class="margin-caption"><strong>Figure 1: Nasopharyngeal vs Nasal, Mid-Turbinate, and combined Oropharyngeal/ Nasal swabs.</strong> Data from Kojima et al.&nbsp;2020, Tu et al.&nbsp;2020, Muller et al.&nbsp;2021, and Desmet et al.&nbsp;2021.</figcaption>
</figure>
</div>
</div>
</div>
<p>All studies, bar one, show equivalent or superior performance of nasopharyngeal swabs. The outlier study, (Tu et al.&nbsp;2020) shows lower CT values for self-collected mid-turbinate swabs vs professionally collected NP swabs. In the same study, nasal swabs prove equivalent to NP swabs. Combined oro-pharyngeal swabs are also equivalent to NP swabs.</p>
<p>Studies that use NP swabs as their gold standard are useful to better understand the performance of other swab types, but NP swabs themselves are not practical for a large-scale pooled sampling programs: NP swabs are notoriously uncomfortable and are commonly administered by a third person. Conversely, nasal or oropharyngeal swabs would be more suitable for self-sampling, which allows far higher testing throughput. Let’s thus look at studies that compare nasal and oropharyngeal swabs.</p>
</section>
<section id="nasal-swabs" class="level3 page-columns page-full">
<h3 class="anchored" data-anchor-id="nasal-swabs">Nasal Swabs</h3>
<p>In qPCR measurements, nasal swabs contain more virus copies than throat swabs. In (Goodall et al.&nbsp;2022), nasal swabs are slightly superior to throat swabs and combined/nasal throat swabs, and come out about even when compared to combined nasal/OP swabs. In (Leung et al.&nbsp;2020), researchers ran multiplex-PCR on both nasal and throat swamples. qPCR differences are plotted in Figure 3 for Human Coronavirus, Influenza, and Rhinovirus.</p>
<div id="681557a2" class="cell page-columns page-full" data-execution_count="3">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="681557a2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="681557a2-1"><a href="#681557a2-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.read_csv(<span class="st">"data/nasal_ct_differences.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="681557a2-2"><a href="#681557a2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-3"><a href="#681557a2-3" aria-hidden="true" tabindex="-1"></a>df.columns <span class="op">=</span> df.columns.<span class="bu">str</span>.replace(<span class="st">", "</span>, <span class="st">"</span><span class="ch">\n</span><span class="st">"</span>)</span>
<span id="681557a2-4"><a href="#681557a2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># reshape dataframe to long format</span></span>
<span id="681557a2-5"><a href="#681557a2-5" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.melt(var_name<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>, value_name<span class="op">=</span><span class="st">"CT Difference"</span>)</span>
<span id="681557a2-6"><a href="#681557a2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-7"><a href="#681557a2-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-8"><a href="#681557a2-8" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="fl">1.5</span>))</span>
<span id="681557a2-9"><a href="#681557a2-9" aria-hidden="true" tabindex="-1"></a>sns.stripplot(</span>
<span id="681557a2-10"><a href="#681557a2-10" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df,</span>
<span id="681557a2-11"><a href="#681557a2-11" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="681557a2-12"><a href="#681557a2-12" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">"CT Difference"</span>,</span>
<span id="681557a2-13"><a href="#681557a2-13" aria-hidden="true" tabindex="-1"></a>    hue<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="681557a2-14"><a href="#681557a2-14" aria-hidden="true" tabindex="-1"></a>    jitter<span class="op">=</span><span class="va">True</span>,</span>
<span id="681557a2-15"><a href="#681557a2-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="681557a2-16"><a href="#681557a2-16" aria-hidden="true" tabindex="-1"></a>plt.legend([], [], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="681557a2-17"><a href="#681557a2-17" aria-hidden="true" tabindex="-1"></a><span class="co"># drop y axis label</span></span>
<span id="681557a2-18"><a href="#681557a2-18" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">""</span>)</span>
<span id="681557a2-19"><a href="#681557a2-19" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"qPCR Cycle Threshold Value Difference"</span>)</span>
<span id="681557a2-20"><a href="#681557a2-20" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">"y"</span>, which<span class="op">=</span><span class="st">"both"</span>, left<span class="op">=</span><span class="va">False</span>, right<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">True</span>)</span>
<span id="681557a2-21"><a href="#681557a2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-22"><a href="#681557a2-22" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> <span class="dv">5</span>, <span class="dv">0</span>, <span class="op">-</span><span class="dv">5</span>, <span class="op">-</span><span class="dv">10</span>, <span class="op">-</span><span class="dv">15</span>:</span>
<span id="681557a2-23"><a href="#681557a2-23" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> x <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="681557a2-24"><a href="#681557a2-24" aria-hidden="true" tabindex="-1"></a>        plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"red"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="681557a2-25"><a href="#681557a2-25" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="681557a2-26"><a href="#681557a2-26" aria-hidden="true" tabindex="-1"></a>    plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="681557a2-27"><a href="#681557a2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-28"><a href="#681557a2-28" aria-hidden="true" tabindex="-1"></a>plt.axhline(y<span class="op">=</span><span class="fl">0.5</span>, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="681557a2-29"><a href="#681557a2-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-30"><a href="#681557a2-30" aria-hidden="true" tabindex="-1"></a>min_x, max_x <span class="op">=</span> plt.xlim()</span>
<span id="681557a2-31"><a href="#681557a2-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-32"><a href="#681557a2-32" aria-hidden="true" tabindex="-1"></a>plt.text(max_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Comparison"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="681557a2-33"><a href="#681557a2-33" aria-hidden="true" tabindex="-1"></a>plt.text(min_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Nasal Swab"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="681557a2-34"><a href="#681557a2-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-35"><a href="#681557a2-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-36"><a href="#681557a2-36" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="681557a2-37"><a href="#681557a2-37" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="681557a2-38"><a href="#681557a2-38" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="681557a2-39"><a href="#681557a2-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="681557a2-40"><a href="#681557a2-40" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full">
<p><img src="swab-sampling_files/figure-html/cell-4-output-1.png" class="figure-img" width="820" height="186"></p>
<figcaption class="margin-caption"><strong>Figure 2: Nasal swabs vs Oro-pharyngeal and combined nasal/oro-pharyngeal swabs.</strong> All data is taken from Goddall et al.&nbsp;2022.</figcaption>
</figure>
</div>
</div>
</div>
<div id="ca252634" class="cell page-columns page-full" data-execution_count="4">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="ca252634"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="ca252634-1"><a href="#ca252634-1" aria-hidden="true" tabindex="-1"></a><span class="co"># All CT difference data was calculated and is taken from https://docs.google.com/spreadsheets/d/1YP4mxT_vxiFwXU5ZuBM4obq_oscfhe05ODWwHFHEznM/</span></span>
<span id="ca252634-2"><a href="#ca252634-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="ca252634-3"><a href="#ca252634-3" aria-hidden="true" tabindex="-1"></a><span class="co"># turn tsv into dataframe. Ignore first row. Second row is column names.</span></span>
<span id="ca252634-4"><a href="#ca252634-4" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.read_csv(<span class="st">"data/leung_genome_copy_differences.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="ca252634-5"><a href="#ca252634-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="ca252634-6"><a href="#ca252634-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="ca252634-7"><a href="#ca252634-7" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> df.melt(var_name<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>, value_name<span class="op">=</span><span class="st">"Genome Copy Number Difference"</span>)</span>
<span id="ca252634-8"><a href="#ca252634-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="ca252634-9"><a href="#ca252634-9" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"Study &amp; Comparison"</span>] <span class="op">=</span> df[<span class="st">"Study &amp; Comparison"</span>].<span class="bu">str</span>.split(<span class="st">","</span>).<span class="bu">str</span>[<span class="op">-</span><span class="dv">1</span>]</span>
<span id="ca252634-10"><a href="#ca252634-10" aria-hidden="true" tabindex="-1"></a>fig <span class="op">=</span> plt.figure(figsize<span class="op">=</span>(<span class="dv">8</span>, <span class="dv">2</span>))</span>
<span id="ca252634-11"><a href="#ca252634-11" aria-hidden="true" tabindex="-1"></a>sns.stripplot(</span>
<span id="ca252634-12"><a href="#ca252634-12" aria-hidden="true" tabindex="-1"></a>    data<span class="op">=</span>df,</span>
<span id="ca252634-13"><a href="#ca252634-13" aria-hidden="true" tabindex="-1"></a>    y<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="ca252634-14"><a href="#ca252634-14" aria-hidden="true" tabindex="-1"></a>    x<span class="op">=</span><span class="st">"Genome Copy Number Difference"</span>,</span>
<span id="ca252634-15"><a href="#ca252634-15" aria-hidden="true" tabindex="-1"></a>    hue<span class="op">=</span><span class="st">"Study &amp; Comparison"</span>,</span>
<span id="ca252634-16"><a href="#ca252634-16" aria-hidden="true" tabindex="-1"></a>    jitter<span class="op">=</span><span class="va">True</span>,</span>
<span id="ca252634-17"><a href="#ca252634-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="ca252634-18"><a href="#ca252634-18" aria-hidden="true" tabindex="-1"></a>plt.legend([], [], frameon<span class="op">=</span><span class="va">False</span>)</span>
<span id="ca252634-19"><a href="#ca252634-19" aria-hidden="true" tabindex="-1"></a><span class="co"># drop y axis label</span></span>
<span id="ca252634-20"><a href="#ca252634-20" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">""</span>)</span>
<span id="ca252634-21"><a href="#ca252634-21" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">"Genome Copy Number Difference (Nasal - Throat Swabs), log10 scale"</span>)</span>
<span id="ca252634-22"><a href="#ca252634-22" aria-hidden="true" tabindex="-1"></a><span class="co"># flip x axis</span></span>
<span id="ca252634-23"><a href="#ca252634-23" aria-hidden="true" tabindex="-1"></a>plt.gca().invert_xaxis()</span>
<span id="ca252634-24"><a href="#ca252634-24" aria-hidden="true" tabindex="-1"></a>plt.tick_params(axis<span class="op">=</span><span class="st">"y"</span>, which<span class="op">=</span><span class="st">"both"</span>, left<span class="op">=</span><span class="va">False</span>, right<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">True</span>)</span>
<span id="ca252634-25"><a href="#ca252634-25" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> x <span class="kw">in</span> <span class="op">-</span><span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">2</span>, <span class="dv">4</span>, <span class="dv">6</span>:</span>
<span id="ca252634-26"><a href="#ca252634-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> x <span class="op">==</span> <span class="dv">0</span>:</span>
<span id="ca252634-27"><a href="#ca252634-27" aria-hidden="true" tabindex="-1"></a>        plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"red"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="ca252634-28"><a href="#ca252634-28" aria-hidden="true" tabindex="-1"></a>        <span class="cf">continue</span></span>
<span id="ca252634-29"><a href="#ca252634-29" aria-hidden="true" tabindex="-1"></a>    plt.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="ca252634-30"><a href="#ca252634-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="ca252634-31"><a href="#ca252634-31" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> y <span class="kw">in</span> <span class="fl">0.5</span>, <span class="fl">1.5</span>:</span>
<span id="ca252634-32"><a href="#ca252634-32" aria-hidden="true" tabindex="-1"></a>    plt.axhline(y<span class="op">=</span>y, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="ca252634-33"><a href="#ca252634-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="ca252634-34"><a href="#ca252634-34" aria-hidden="true" tabindex="-1"></a>min_x, max_x <span class="op">=</span> plt.xlim()</span>
<span id="ca252634-35"><a href="#ca252634-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="ca252634-36"><a href="#ca252634-36" aria-hidden="true" tabindex="-1"></a>plt.text(max_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Throat Swab"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="ca252634-37"><a href="#ca252634-37" aria-hidden="true" tabindex="-1"></a>plt.text(min_x <span class="op">/</span> <span class="dv">2</span>, <span class="op">-</span><span class="fl">0.6</span>, <span class="st">"Favors Nasal Swab"</span>, fontsize<span class="op">=</span><span class="dv">10</span>, color<span class="op">=</span><span class="st">"black"</span>, ha<span class="op">=</span><span class="st">"center"</span>)</span>
<span id="ca252634-38"><a href="#ca252634-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="ca252634-39"><a href="#ca252634-39" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="ca252634-40"><a href="#ca252634-40" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="ca252634-41"><a href="#ca252634-41" aria-hidden="true" tabindex="-1"></a>plt.gca().spines[<span class="st">"left"</span>].set_visible(<span class="va">False</span>)</span>
<span id="ca252634-42"><a href="#ca252634-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="ca252634-43"><a href="#ca252634-43" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full">
<p><img src="swab-sampling_files/figure-html/cell-5-output-1.png" class="figure-img" width="760" height="223"></p>
<figcaption class="margin-caption"><strong>Figure 3: Nasal swabs vs Oro-pharyngeal swabs.</strong> All data is taken from Leung et al.&nbsp;2020.</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="virus-relative-abundance-across-different-swab-types" class="level1 page-columns page-full">
<h1>Virus relative abundance across different swab types</h1>
<p>Whatever the sample type, ultimately we will want to perform metagenomic sequencing to detect novel pathogens. As described in a <a href="https://naobservatory.org/reports/comparing-sampling-strategies-for-early-detection-of-stealth-biothreats/">previous report</a>, the value of a metagenomic sequencing sample will can be assessed across several dimensions:</p>
<ul>
<li>What is the expected relative abundance of a human-infecting stealth virus? </li>
<li>How complex are individual sequencing samples? I.e, what is their species richness and evenness?</li>
<li>How much does the background complexity of a sample vary between individuals and over time?</li>
</ul>
<div class="no-row-height column-margin column-container"><span class="margin-aside">The shedding dynamics of a <strong>stealth pathogen</strong> are likely to be different from SARS-CoV-2, so we need to be careful about generalizing from samples that originate from SARS-CoV-2 patients, particularly if they are hospitalized.</span></div><p>Three studies performed paired metagenomic sequencing and SARS-CoV-2 qPCR on swab samples, Castañeda-Mogollón et al.&nbsp;2021, Lu et al.&nbsp;2021, Babiker et al.&nbsp;2020. In their results Castañeda-Mogollón et al.&nbsp;2021 only shows absolute SARS-CoV-2 genome copies, so we will skip this study for now.</p>
<div id="5dc5ce37" class="cell page-columns page-full" data-execution_count="5">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="5dc5ce37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="5dc5ce37-1"><a href="#5dc5ce37-1" aria-hidden="true" tabindex="-1"></a>df_throat <span class="op">=</span> pd.read_csv(<span class="st">"data/lu_throat_ct_mgs.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="5dc5ce37-2"><a href="#5dc5ce37-2" aria-hidden="true" tabindex="-1"></a>df_throat.rename(</span>
<span id="5dc5ce37-3"><a href="#5dc5ce37-3" aria-hidden="true" tabindex="-1"></a>    columns<span class="op">=</span>{</span>
<span id="5dc5ce37-4"><a href="#5dc5ce37-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SARS-CoV-2 RT-PCR Ct"</span>: <span class="st">"scv2_ct"</span>,</span>
<span id="5dc5ce37-5"><a href="#5dc5ce37-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"SARS-CoV-2 RA"</span>: <span class="st">"scv2_ra"</span>,</span>
<span id="5dc5ce37-6"><a href="#5dc5ce37-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Inpatient/ED vs. Outpatient"</span>: <span class="st">"patient_status"</span>,</span>
<span id="5dc5ce37-7"><a href="#5dc5ce37-7" aria-hidden="true" tabindex="-1"></a>    },</span>
<span id="5dc5ce37-8"><a href="#5dc5ce37-8" aria-hidden="true" tabindex="-1"></a>    inplace<span class="op">=</span><span class="va">True</span>,</span>
<span id="5dc5ce37-9"><a href="#5dc5ce37-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="5dc5ce37-10"><a href="#5dc5ce37-10" aria-hidden="true" tabindex="-1"></a>df_throat[<span class="st">"scv2_ct"</span>] <span class="op">=</span> df_throat[<span class="st">"scv2_ct"</span>].replace(<span class="st">","</span>, <span class="st">"."</span>, regex<span class="op">=</span><span class="va">True</span>).astype(<span class="bu">float</span>)</span>
<span id="5dc5ce37-11"><a href="#5dc5ce37-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="5dc5ce37-12"><a href="#5dc5ce37-12" aria-hidden="true" tabindex="-1"></a>df_nasopharyngeal <span class="op">=</span> pd.read_csv(<span class="st">"data/babiker_np_ct_mgs.tsv"</span>, sep<span class="op">=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, skiprows<span class="op">=</span><span class="dv">1</span>)</span>
<span id="5dc5ce37-13"><a href="#5dc5ce37-13" aria-hidden="true" tabindex="-1"></a>df_nasopharyngeal.rename(</span>
<span id="5dc5ce37-14"><a href="#5dc5ce37-14" aria-hidden="true" tabindex="-1"></a>    columns<span class="op">=</span>{<span class="st">"SCV-2 Relative Abundance"</span>: <span class="st">"scv2_ra"</span>, <span class="st">"Ct value"</span>: <span class="st">"scv2_ct"</span>}, inplace<span class="op">=</span><span class="va">True</span></span>
<span id="5dc5ce37-15"><a href="#5dc5ce37-15" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="5dc5ce37-16"><a href="#5dc5ce37-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="5dc5ce37-17"><a href="#5dc5ce37-17" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">2</span>, <span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">9</span>, <span class="dv">6</span>))</span>
<span id="5dc5ce37-18"><a href="#5dc5ce37-18" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> i, (ax, df) <span class="kw">in</span> <span class="bu">enumerate</span>(<span class="bu">zip</span>(axs, [df_nasopharyngeal, df_throat])):</span>
<span id="5dc5ce37-19"><a href="#5dc5ce37-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> i <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="5dc5ce37-20"><a href="#5dc5ce37-20" aria-hidden="true" tabindex="-1"></a>        <span class="co"># set dot colors</span></span>
<span id="5dc5ce37-21"><a href="#5dc5ce37-21" aria-hidden="true" tabindex="-1"></a>        sns_default_colors <span class="op">=</span> sns.color_palette(<span class="st">"tab10"</span>)</span>
<span id="5dc5ce37-22"><a href="#5dc5ce37-22" aria-hidden="true" tabindex="-1"></a>        sns.scatterplot(</span>
<span id="5dc5ce37-23"><a href="#5dc5ce37-23" aria-hidden="true" tabindex="-1"></a>            ax<span class="op">=</span>ax,</span>
<span id="5dc5ce37-24"><a href="#5dc5ce37-24" aria-hidden="true" tabindex="-1"></a>            data<span class="op">=</span>df,</span>
<span id="5dc5ce37-25"><a href="#5dc5ce37-25" aria-hidden="true" tabindex="-1"></a>            x<span class="op">=</span><span class="st">"scv2_ct"</span>,</span>
<span id="5dc5ce37-26"><a href="#5dc5ce37-26" aria-hidden="true" tabindex="-1"></a>            y<span class="op">=</span><span class="st">"scv2_ra"</span>,</span>
<span id="5dc5ce37-27"><a href="#5dc5ce37-27" aria-hidden="true" tabindex="-1"></a>            hue<span class="op">=</span><span class="st">"patient_status"</span>,</span>
<span id="5dc5ce37-28"><a href="#5dc5ce37-28" aria-hidden="true" tabindex="-1"></a>            style<span class="op">=</span><span class="st">"patient_status"</span>,</span>
<span id="5dc5ce37-29"><a href="#5dc5ce37-29" aria-hidden="true" tabindex="-1"></a>            s<span class="op">=</span><span class="dv">100</span>,</span>
<span id="5dc5ce37-30"><a href="#5dc5ce37-30" aria-hidden="true" tabindex="-1"></a>            palette<span class="op">=</span>sns_default_colors[<span class="dv">2</span>:<span class="dv">4</span>],</span>
<span id="5dc5ce37-31"><a href="#5dc5ce37-31" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="5dc5ce37-32"><a href="#5dc5ce37-32" aria-hidden="true" tabindex="-1"></a>        <span class="co"># set coordinates for legend</span></span>
<span id="5dc5ce37-33"><a href="#5dc5ce37-33" aria-hidden="true" tabindex="-1"></a>        coordinates <span class="op">=</span> (<span class="fl">0.6</span>, <span class="op">-</span><span class="fl">0.2</span>)</span>
<span id="5dc5ce37-34"><a href="#5dc5ce37-34" aria-hidden="true" tabindex="-1"></a>        ax.legend(title<span class="op">=</span><span class="st">""</span>, loc<span class="op">=</span><span class="st">"upper left"</span>)  <span class="co"># bbox_to_anchor=coordinates)</span></span>
<span id="5dc5ce37-35"><a href="#5dc5ce37-35" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">"SARS-CoV-2 qPCR CT Value"</span>)</span>
<span id="5dc5ce37-36"><a href="#5dc5ce37-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="5dc5ce37-37"><a href="#5dc5ce37-37" aria-hidden="true" tabindex="-1"></a>    <span class="cf">else</span>:</span>
<span id="5dc5ce37-38"><a href="#5dc5ce37-38" aria-hidden="true" tabindex="-1"></a>        sns.scatterplot(ax<span class="op">=</span>ax, data<span class="op">=</span>df, x<span class="op">=</span><span class="st">"scv2_ct"</span>, y<span class="op">=</span><span class="st">"scv2_ra"</span>, s<span class="op">=</span><span class="dv">100</span>)</span>
<span id="5dc5ce37-39"><a href="#5dc5ce37-39" aria-hidden="true" tabindex="-1"></a>        ax.set_xlabel(<span class="st">""</span>)</span>
<span id="5dc5ce37-40"><a href="#5dc5ce37-40" aria-hidden="true" tabindex="-1"></a>    ax.set_yscale(<span class="st">"log"</span>)</span>
<span id="5dc5ce37-41"><a href="#5dc5ce37-41" aria-hidden="true" tabindex="-1"></a>    ax.invert_xaxis()</span>
<span id="5dc5ce37-42"><a href="#5dc5ce37-42" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"right"</span>].set_visible(<span class="va">False</span>)</span>
<span id="5dc5ce37-43"><a href="#5dc5ce37-43" aria-hidden="true" tabindex="-1"></a>    ax.spines[<span class="st">"top"</span>].set_visible(<span class="va">False</span>)</span>
<span id="5dc5ce37-44"><a href="#5dc5ce37-44" aria-hidden="true" tabindex="-1"></a>    ax.tick_params(axis<span class="op">=</span><span class="st">"y"</span>, which<span class="op">=</span><span class="st">"both"</span>, left<span class="op">=</span><span class="va">False</span>, right<span class="op">=</span><span class="va">False</span>, labelleft<span class="op">=</span><span class="va">True</span>)</span>
<span id="5dc5ce37-45"><a href="#5dc5ce37-45" aria-hidden="true" tabindex="-1"></a>    ax.set_xlim(<span class="fl">37.5</span>, <span class="dv">15</span>)</span>
<span id="5dc5ce37-46"><a href="#5dc5ce37-46" aria-hidden="true" tabindex="-1"></a>    ax.set_ylim(<span class="dv">10</span><span class="op">**-</span><span class="dv">7</span>, <span class="fl">0.5</span>)</span>
<span id="5dc5ce37-47"><a href="#5dc5ce37-47" aria-hidden="true" tabindex="-1"></a>    ax.set_ylabel(<span class="st">"SARS-CoV-2 Relative Abundance"</span>)</span>
<span id="5dc5ce37-48"><a href="#5dc5ce37-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> x <span class="kw">in</span> np.arange(<span class="dv">15</span>, <span class="fl">37.5</span>, <span class="fl">2.5</span>):</span>
<span id="5dc5ce37-49"><a href="#5dc5ce37-49" aria-hidden="true" tabindex="-1"></a>        ax.axvline(x<span class="op">=</span>x, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="5dc5ce37-50"><a href="#5dc5ce37-50" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> y <span class="kw">in</span> <span class="bu">range</span>(<span class="op">-</span><span class="dv">7</span>, <span class="dv">1</span>, <span class="dv">1</span>):</span>
<span id="5dc5ce37-51"><a href="#5dc5ce37-51" aria-hidden="true" tabindex="-1"></a>        log_y <span class="op">=</span> <span class="dv">10</span><span class="op">**</span>y</span>
<span id="5dc5ce37-52"><a href="#5dc5ce37-52" aria-hidden="true" tabindex="-1"></a>        ax.axhline(y<span class="op">=</span>log_y, color<span class="op">=</span><span class="st">"grey"</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, alpha<span class="op">=</span><span class="fl">0.5</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display page-columns page-full">
<div class="quarto-figure quarto-figure-center page-columns page-full">
<figure class="figure page-columns page-full">
<p><img src="swab-sampling_files/figure-html/cell-6-output-1.png" class="figure-img" width="763" height="503"></p>
<figcaption class="margin-caption"><strong>Figure 4: SARS-CoV-2 qPCR CT vs SARS-CoV-2 RA.</strong> Top Figure: Throat Swabs. Data from Lu et al.&nbsp;2021. Bottom Figure: Nasopharyngeal Swabs. Data from Babiker et al.&nbsp;2020.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="supplement" class="level1">
<h1>Supplement</h1>
<p>Create a figure for the remaining viruses not covered in Figure 3.</p>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>